app/src/main/java/net/aiscope/gdd_app/ui/mask/customview/MaskLayer.kt

```diff
@@ -46,171 +35,225 @@ class MaskLayer(
             this.strokeJoin = Paint.Join.ROUND
             this.strokeCap = Paint.Cap.ROUND
             this.color = color
-            this.alpha = MASK_PAINT_ALPHA
+            this.alpha = (ALPHA_OPAQUE * MASK_PAINT_ALPHA).toInt()
+            this.strokeWidth = strokeWidth
+        }
+
+        fun newDefaultPaintEraser(strokeWidth: Float) = Paint().apply {
+            this.xfermode = ERASER_XFER_MODE
+            this.isAntiAlias = true
+            this.isDither = true
+            this.style = Paint.Style.STROKE
+            this.strokeJoin = Paint.Join.ROUND
+            this.strokeCap = Paint.Cap.ROUND
             this.strokeWidth = strokeWidth
         }
     }
 
-    private var pathsPaintsAndStagesNames: MutableList<PathPaintAndStageName> = LinkedList()
-    private var undoPendingPaths = 0
-    private lateinit var currentPath: PointToPointPath
-    private var currentPaintDirty = false
-    var brushDiseaseStage = BrushDiseaseStage(0, "non-initialized", 0)
+    //fields lazily initialized
+    private var _width: Int? = null
+    private var _height: Int? = null
+    var currentScale = 1f
         set(value) {
-            currentPaintDirty = currentPaintDirty || field != value
             field = value
+            paintBrushPendingRecreation = true
+            paintEraserPendingRecreation = true
         }
-    private var currentStrokeWidth = PATH_STROKE
+
+    //fields depending on init of dimensions
+    private val width by lazy { _width!! }
+    private val height by lazy { _height!! }
+    private val latestChangeBitmap by lazy {
+        Bitmap.createBitmap(
+            width,
+            height,
+            Bitmap.Config.ARGB_8888
+        )
+    }
+    private val latestChangeBitmapCanvas by lazy { Canvas(latestChangeBitmap) }
+    private val currentStateBitmap by lazy {
+        Bitmap.createBitmap(
+            width,
+            height,
+            Bitmap.Config.ARGB_8888
+        )
+    }
+    private val currentStateBitmapCanvas by lazy { Canvas(currentStateBitmap) }
+
+    //fields depending on init of stage
+    private lateinit var currentDiseaseStage: BrushDiseaseStage
+    private lateinit var currentPaintBrush: Paint
+    private lateinit var currentPaintEraser: Paint
+    private lateinit var currentPaint: Paint
+
+    //init independent fields
+    private var paintBrushPendingRecreation = true
+    private var paintEraserPendingRecreation = true
+    private var pathsAndPaints: MutableList<PathAndPaint> = LinkedList()
+    private var undoPendingPaths = 0
+    private var currentPath: PointToPointPath? = null
```

Comment:
+1 to this

Reply:
Because we're relying on it being non-null just when we're between a touch-start and a touch-end... That's used as a condition somewhere else to determine whether we're currently drawing something or not. The point I can remember this happens is when we rotate the screen, for instance.

Reply:
I'm not generally a big fan of using `null` as an indicator of state (In this case "not-drawing")

Should we just have an active `Mode`, same as in `MaskCustomView`?

Also, if we _are_ going to let this be nullable, I'd far prefer if the handling of this is in safe calls (`?.`) or elvis (`?:`) rather than "We know this can never be null" (`!!`)

Reply:
> Should we just have an active Mode, same as in MaskCustomView?
Would a boolean "currentlyDrawing" suffice?

> Also, if we are going to let this be nullable, I'd far prefer if the handling of this is in safe calls (?.) or elvis (?:) rather than "We know this can never be null" (!!)
I'd like to discuss this further when we can - my main doubt here is: would not that approach possibly mask an illegal state? How can we take advantage of safe calls in case where we need the object to be set?

Reply:
> Would a boolean "currentlyDrawing" suffice?
Well that assumes there can only ever be 2 states: drawing and not drawing. But hey, I can live with that

Reply:
> would not that approach possibly mask an illegal state?

So you would prefer to fail hard than fail silent is what you're saying? Dunno man. @eduardb any opinion?

Reply:
Generally speaking, I prefer failing hard than failing silently and end up in an inconsistent state. We have Crashlytics, and any crashes will be reported, so we can fix them. The downside is that we'll have to deploy a new version with the fix.
With all this being said, I think that in this case, it's still possible to model the problem space better and avoid the double-bang operator. I will have a closer look alter to see if I can come up with something.

Reply:
@eduardb created a PR with his suggestions, and we got it merged. Thanks!!